ouroboros-consensus-cardano/test/Test/ThreadNet/TxGen/Cardano.hs

```diff
@@ -1,15 +1,308 @@
+          , SL._mdHash   = SL.SNothing
+          , SL._outputs  =
+              StrictSeq.singleton $ SL.TxOut shelleyAddr unspentCoin
+          , SL._ttl      = SlotNo maxBound
+          , SL._txUpdate = SL.SNothing
+          , SL._txfee    = fee
+          , SL._wdrls    = SL.Wdrl Map.empty
+          }
+
+        bodyHash :: SL.Hash c (SL.TxBody c)
+        bodyHash = SL.hashWithSerialiser toCBOR body
+
+        -- Witness the use of bootstrap address's utxo.
+        byronWit :: SL.BootstrapWitness c
+        byronWit =
+            SL.makeBootstrapWitness bodyHash byronSK $
+            Byron.addrAttributes byronAddr
+
+        -- Witness the stake delegation.
+        delegWit :: SL.WitVKey c 'SL.Witness
+        delegWit =
+            SL.WitVKey
+              (Shelley.mkVerKey stakingSK)
+              (SL.signedDSIGN @c stakingSK bodyHash)
+
+        -- Witness the pool registration.
+        poolWit :: SL.WitVKey c 'SL.Witness
+        poolWit =
+            SL.WitVKey
+              (Shelley.mkVerKey poolSK)
+              (SL.signedDSIGN @c poolSK bodyHash)
+
+    in
+    if Map.null picked then Nothing else
+    (Just . GenTxShelley. mkShelleyTx) $
+    SL.Tx
+      { SL._body       = body
+      , SL._metadata   = SL.SNothing
+      , SL._witnessSet = SL.WitnessSet
+          { SL.addrWits = Set.fromList [delegWit, poolWit]
+          , SL.bootWits = Set.singleton byronWit
+          , SL.msigWits = Map.empty
+          }
+      }
+
+    | otherwise           = Nothing
+
+  where
+    mbUTxO :: Maybe (SL.UTxO c)
+    mbUTxO =
+        fmap getUTxOShelley $
+        ejectShelleyTickedLedgerState $
+        applyChainTick lcfg curSlot $
+        lst
+
+    MigrationInfo
+      { byronMagic
+      , byronSK
+      , paymentSK
+      , poolSK
+      , stakingSK
+      , vrfSK
+      } = migrationInfo
+
+    byronAddr :: Byron.Address
+    byronAddr =
+        Byron.makeVerKeyAddress byronMagic $ toVerification byronSK
+
+    -- We use a base reference for the stake so that we can refer to it in the
+    -- same tx that registers it.
+    shelleyAddr :: SL.Addr c
+    shelleyAddr =
+        SL.Addr Shelley.networkId
+          (Shelley.mkCredential paymentSK)
+          (SL.StakeRefBase $ Shelley.mkCredential stakingSK)
+
+    -- A simplistic individual pool
+    poolParams :: SL.Coin -> SL.PoolParams c
+    poolParams pledge = SL.PoolParams
+        { SL._poolCost   = SL.Coin 1
+        , SL._poolMD     = SL.SNothing
+        , SL._poolMargin = SL.truncateUnitInterval 0
+        , SL._poolOwners = Set.singleton $ Shelley.mkKeyHash poolSK
+        , SL._poolPledge = pledge
+        , SL._poolPubKey = Shelley.mkKeyHash poolSK
+        , SL._poolRAcnt  =
+            SL.RewardAcnt Shelley.networkId $ Shelley.mkCredential poolSK
+        , SL._poolRelays = StrictSeq.empty
+        , SL._poolVrf    = Shelley.mkKeyHashVrf vrfSK
+        }
+
+-----
+
+ejectShelleyNS :: NS f (CardanoEras c) -> Maybe (f (ShelleyBlock c))
```

Comment:
@mrBliss Are there pre-existing functions that replace/simplify these `ejectShelleyNS`, `ejectShelleyTickedLedgerState`, and `getUTxOShelley` functions I'm defining here?

Reply:
There's is the [`LedgerStateShelley`](https://github.com/input-output-hk/ouroboros-network/blob/3a83c64ce76fb995b9557d3eb55d3c7e1c2fc251/ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs#L494) pattern synonym, but that's not for a *ticked* ledger state, and I don't want to add such a synonym as I don't want other packages to mess with ticked ledger states.

No, I don't think there are any pre-existing functions you can use to replace/simplify these three.

Why "eject"? Is that an existing naming convention? It makes me think of "reject", which would do the opposite (so in some sense, "eject" makes sense). I would go for something like "select" instead.

Reply:
We agreed to follow https://hackage.haskell.org/package/generics-sop-0.5.1.0/docs/Generics-SOP.html#t:Ejection